### ISSUE TITLE
NBC - rationale against MPI_Request_free

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#83]
**Originally by _htor_ on 2008-12-29 09:30:33 -0600**

---

Chapter 5.12 (draft page 50:19-21): 

"Rationale. Freeing an active nonblocking collective request could cause similar problems as discussed for point-to-point requests (see Section 3.7.3)."

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2008-12-31 10:57:57 -0600**

---

I would suggest some minor rewording to make the text clearer. Specifically, change it to:

Rationale. Freeing a nonblocking collective request is erroneous because all collectives include a receive action for at least some of the processes. Thus, freeing these requests is erroneous since there is no portable way to verify that the operation is completed and the receive buffer can be used after an active receive request is freed (see Section 3.7.3).

Having added this rewording, it seems like the rule is overly restrictive. We could easily allow freeing of nonblocking broadcast, although the logic of the 
program would require that the user knows all processes have received the data. this could be accomplished by a barrier that follows the completion on all other processes, for example. Nonetheless, it seems a minor restriction so I will not lobby against it.

I have a few other minor wording changes to suggest that are related to the rationale to which this issue relates. First, the current wording right before the rationale is not the correct wording for the standard. Also, the wording is too unspecific. Currently, page 50, line 16 reads "MPI_REQUEST_FREE is not supported. MPI_CANCEL is not supported." The term that should be used is "erroneous" so I suggest that these sentences be changed to "Calling MPI_REQUEST_FREE or MPI_CANCEL for a nonblocking collective request is erroneous." This also makes it clear that the statement only applies to collectives.

A much smaller change is for page 50, line 22, which has "can not" as two words; "cannot" is only one word in correct English.

Finally, I find the rationale for not having tags too colloquial. Currently, page 50, lines 25-26 read:

Not having tags for collective operations enables certain hardware-optimizations and is consistent with blocking collective operations. (End of rationale.)

I suggest changing these lines to the following:

The use of tags for collective operations would prevent certain hardware-optimizations and is consistent with blocking collective operations. (End of rationale.)

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2008-12-31 11:24:16 -0600**

---

Replying to bronis:

Thanks for your review, let me comment on your comments:

> I would suggest some minor rewording to make the text clearer. Specifically, change it to:
> 
> Rationale. Freeing a nonblocking collective request is erroneous because all collectives include a receive action for at least some of the processes. Thus, freeing these requests is erroneous since there is no portable way to verify that the operation is completed and the receive buffer can be used after an active receive request is freed (see Section 3.7.3).
> I agree - I changed this in the document and the proposal text above!
> 
> Having added this rewording, it seems like the rule is overly restrictive. We could easily allow freeing of nonblocking broadcast, although the logic of the program would require that the user knows all processes have received the data. this could be accomplished by a barrier that follows the completion on all other processes, for example. Nonetheless, it seems a minor restriction so I will not lobby against it.
> I think you argue towards using MPI_Request_free of a broadcast request on the root process only, right? This use-case seems rather restricted to me and unless we have a better use-case, I would argue against it. I could be convinced with a good use-case though, but we agreed that freeing a request is problematic (and has all kinds of ugly side effects, e.g., on MPI_THREAD_MULTIPLE implementations (another beer topic)) so that I would not encourage its use and slowly deprecate it. A barrier only ensures that all processes have called the operation, it is neither a communication barrier nor a memory barrier. Thus, it only guarantees that all processes started the nonblocking operation which was called before the barrier. No guarantee about the data consistency can be assumed. Furthermore, all the discussions related to #47 apply also to nonblocking collectives. I cc'd Hubert to voice his opinions on Request_free.
> 
> I have a few other minor wording changes to suggest that are related to the rationale to which this issue relates. First, the current wording right before the rationale is not the correct wording for the standard. Also, the wording is too unspecific. Currently, page 50, line 16 reads "MPI_REQUEST_FREE is not supported. MPI_CANCEL is not supported." The term that should be used is "erroneous" so I suggest that these sentences be changed to "Calling MPI_REQUEST_FREE or MPI_CANCEL for a nonblocking collective request is erroneous." This also makes it clear that the statement only applies to collectives.
> yes, I agree, I changed your sentence slightly to:
> "Calling MPI_REQUEST_FREE or MPI_CANCEL for a request associated with a nonblocking collective is erroneous."
> 
> A much smaller change is for page 50, line 22, which has "can not" as two words; "cannot" is only one word in correct English.
> thanks! Corrected it.
> 
> Finally, I find the rationale for not having tags too colloquial. Currently, page 50, lines 25-26 read:
> 
> Not having tags for collective operations enables certain hardware-optimizations and is consistent with blocking collective operations. (End of rationale.)
> 
> I suggest changing these lines to the following:
> 
> The use of tags for collective operations would prevent certain hardware-optimizations and is consistent with blocking collective operations. (End of rationale.)
> I agree and changed the proposal!

I updated the document, the current draft is now: https://svn.mpi-forum.org/trac/mpi-forum-web/attachment/wiki/NBColl/nbc-proposal-12-31-2008-2.pdf

**Comment by mpiforumbot** [comment]
**Originally by _hubertritzdorf_ on 2009-01-12 11:24:16 -0600**

---

I think that this rationale doesn't make sense. MPI_REQUEST_FREE is allowed for non-blocking (pt2pt) receive requests nethertheless the same arguments against MPI_REQUEST_FREE would be true for non-blocking (pt2pt) receives.

MPI_REQUEST_FREE is provided for convenience on the sending side.
Therefore, I would simply allow the usage of MPI_REQUEST_FREE with the following reasons:
  (_) Completeness[[BR]]
  (_) Functions without send and receive buffers:
      Freeing of requests in case of MPI_IBarrier (for example: applications where
      slave processes want to inform a master process)
  (_) Requests for which only a send buffer is provided:
      Freeing of requests where only a send buffer is provided (IGather, IBcast, ...)
  (_) Requests where data is not received:
      Freeing of requests where the recv count is 0
      (Iallgatherv, Iscatterv, ...)

**Comment by mpiforumbot** [comment]
**Originally by _erezh_ on 2009-01-12 12:35:17 -0600**

---

I'm not sure that you're going to need this rationale; we're going to have a proposal to make MPI_Request_free erroneous for active requests. that is any active request including nonblocking collectives request.

Makes sense?

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-12 13:27:55 -0600**

---

Replying to hubertritzdorf:

> I think that this rationale doesn't make sense. MPI_REQUEST_FREE is allowed for non-blocking (pt2pt) receive requests nethertheless the same arguments against MPI_REQUEST_FREE would be true for non-blocking (pt2pt) receives.
> yes, and this is why we have an advice to users to never use it :). Just because we couldn't change the semantics in MPI-2.1. However, it was shown that there is no portable/reliable/fast way to check for completion.
> 
> MPI_REQUEST_FREE is provided for convenience on the sending side.
> Therefore, I would simply allow the usage of MPI_REQUEST_FREE with the following reasons:
>   (_) Completeness[[BR]]
>   (_) Functions without send and receive buffers:
>       Freeing of requests in case of MPI_IBarrier (for example: applications where
>       slave processes want to inform a master process)
>   (_) Requests for which only a send buffer is provided:
>       Freeing of requests where only a send buffer is provided (IGather, IBcast, ...)
>   (_) Requests where data is not received:
>       Freeing of requests where the recv count is 0
>       (Iallgatherv, Iscatterv, ...) 
> Those are all very valid arguments. Thus, I could be convinced. Request_free would also not be more problematic as it already is in the p2p case with my current implementation. We have to think about what happens for hardware-accelerated implementations. We will discuss this during our telecon this week.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-15 14:12:17 -0600**

---

replaced rationale in current document with "Rationale. Freeing an active nonblocking collective request could cause similar problems as discussed for point-to-point requests (see Section 3.7.3)."!

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-21 06:58:02 -0600**

---

I checked the text based on pdf from Jan 16, 2009.

About MPI_REQUEST_FREE:

  Review: The text is okay.

  My opinion: Never allow MPI_REQUEST_FREE with non-blocking collectives.
  If somebody wants to allow it, then only after significant experience 
  with existing nonblocking collectives in MPI libraries and applications,
  i.e., not before MPI-3.1.

About tags and nonblocking collectives at p50:20 and p50:26-28 of your proposal 
(pdf, Jan 16, 2009):

  Review: p2:30 of your proposal (pdf, Jan 16, 2009) states already
  "The collective operations do not have a message tag argument."
  There isn't a rationale.
  On p50:20, you repeat this sentence and you add a rationale on p50:26-28
  that is vlid only for nonblocking collectives.

  Therefore, the wording looks a little bit inconsistent.

  My proposal 1: Remove p50:20 and p50:26-28.

  My proposal 2:
  Change p50:20 to: "As with blocking collective operations, also 
  nonblocking collective operations do not have a message tag argument."
  and do not put the sentence in a new paragraph.[[BR]]
  Change p50:26-28 to:  
  "The use of tags for nonblocking collectives would prevent certain 
  hardware-optimizations.

Further problems with the current text:

Matching rule on p50:34-45:

  Up to this point, it is not clear, that IBARRIER and BARRIER are different
  collective **operations** (they are different routines, that is clear).
  Therefore, we need early wording telling this.
  I propose to add on p50:35 before "In particular":[[BR]]
  "Blocking and nonblocking collective routines are regarded as 
  different operations." [[BR]]
  and I would remove "(blocking and nonblocking)" on line p50:34.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-21 12:55:01 -0600**

---

Replying to RolfRabenseifner:

> I checked the text based on pdf from Jan 16, 2009.
> 
> About MPI_REQUEST_FREE:
> 
>   Review: The text is okay.
> 
>   My opinion: Never allow MPI_REQUEST_FREE with non-blocking collectives.
>   If somebody wants to allow it, then only after significant experience 
>   with existing nonblocking collectives in MPI libraries and applications,
>   i.e., not before MPI-3.1.
> Yes, I agree - we should wait and see how #47 goes
> 
> About tags and nonblocking collectives at p50:20 and p50:26-28 of your proposal 
> (pdf, Jan 16, 2009):
> I will answer these remarks in ticket #109 (Nonblocking Collectives)!

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-01-22 03:21:02 -0600**

---

Now, all is okay. You addressed all my review items here or in ticket #109.
For me, this ticket can be closed if no other reviews show up.

**Comment by mpiforumbot** [comment]
**Originally by _traff_ on 2009-01-27 08:10:07 -0600**

---

Even though this a minority opinion, I would still suggest allowing MPI_Request_free for 
nonblocking collectives. The issue is more general, and there is no good technical reason 
for forbidding it for the nonblocking colls.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-27 09:14:37 -0600**

---

Replying to traff:

> Even though this a minority opinion, I would still suggest allowing MPI_Request_free for 
> nonblocking collectives. The issue is more general, and there is no good technical reason 
> for forbidding it for the nonblocking colls.
> I would argue that there are good technical reasons. One of them is that it seems impossible to write a portable program that uses the data of, for example, MPI_Ialltoall and frees the active request. Those issues are very similar to #47. I think the issue that request free provokes non-portable programming and dangers related to #47 are a strong technical reason. 

However, if we would allow MPI_Request_free, then we would have to say clearly that nothing can be guaranteed for the consistency of data in the receive buffer and if any guarantees are needed for the reuse of the send buffer, then explicit synchronization is necessary (similar to what we have in the p2p chapter in MPI-2.1). From a clearly technical standpoint, I agree that all of Hubert's examples from above are correct, but they cover a very very small portion of the possibilities and all other uses would be dangerous (mostly for portability).

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-01-27 19:46:26 -0600**

---

this was merged into revision 1 - all discussions are moved to #109 and #47 respectively.
